jena-core/src/main/java/org/apache/jena/graph/impl/LiteralLabelImpl.java

```diff
@@ -303,20 +304,59 @@ public String getLexicalForm() {
 		return lexicalForm;
 	}
     
-    /** 
-     	Answer the value used to index this literal
-        TODO Consider pushing indexing decisions down to the datatype
-    */
+    /**
+     * Answer an object used to index this literal. This object must provide
+     * {@link Object#equals} and {@link Object#hashCode} based on values, not object
+     * instance identity.
+     */
     @Override
     public Object getIndexingValue() {
-        return
-            isXML() ? this
-            : !lang.equals( "" ) ? getLexicalForm() + "@" + lang.toLowerCase(Locale.ROOT)
-            : wellformed ? getValue()
-            : getLexicalForm() 
-            ;
+        if ( isXML() )
+            return this;
+        if ( !lang.equals("") )
+            return getLexicalForm() + "@" + lang.toLowerCase(Locale.ROOT);
+        if ( wellformed ) {
+            Object value = getValue();
+            // JENA-1936
+            // byte[] does not provide hashCode/equals based on the contents of the array.
+            if ( value instanceof byte[] )
+                return new ByteArray((byte[])value);
+            return value;
+        }
+        return getLexicalForm();
     }
 
+    /**
+     * {@code byte[]} wrapper that provides {@code hashCode} and {@code equals} based
+     * on the value of the array.
+     */
+    static class ByteArray {
+        private final byte[] bytes;
+        /*package*/ ByteArray(byte[] bytes) {
```

Comment:
Consider caching the hash

```
private int hash;

/*package*/ ByteArray(byte[] bytes) {
  this.bytes = bytes;
}

@Override
        public int hashCode() {
             if (hash == 0) {
               hash = Arrays..hashCode(bytes);
             }
             return hash;
        }
```

Reply:
OK - done. We are assuming the `byte[]` isn't changing.